utils/staking-miner/src/monitor.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The monitor command.
+
+use crate::{
+	params, prelude::*, rpc_helpers::*, signer::Signer, Error, MonitorConfig, SharedConfig,
+};
+use codec::Encode;
+use jsonrpsee_ws_client::{
+	traits::SubscriptionClient, v2::params::JsonRpcParams, Subscription, WsClient,
+};
+use sp_transaction_pool::TransactionStatus;
+
+/// Ensure that now is the signed phase.
+async fn ensure_signed_phase<T: EPM::Config, B: BlockT>(
+	client: &WsClient,
+	at: B::Hash,
+) -> Result<(), Error> {
+	let key = sp_core::storage::StorageKey(EPM::CurrentPhase::<T>::hashed_key().to_vec());
+	let phase = get_storage::<EPM::Phase<BlockNumber>>(client, params! {key, at})
+		.await?
+		.unwrap_or_default();
+
+	if phase.is_signed() {
+		Ok(())
+	} else {
+		Err(Error::IncorrectPhase)
+	}
+}
+
+/// Ensure that our current `us` have not submitted anything previously.
+async fn ensure_no_previous_solution<
+	T: EPM::Config + frame_system::Config<AccountId = AccountId>,
+	B: BlockT,
+>(
+	ext: &mut Ext,
+	us: &AccountId,
+) -> Result<(), Error> {
+	use EPM::signed::SignedSubmissions;
+	ext.execute_with(|| {
+		if <SignedSubmissions<T>>::get().iter().any(|ss| &ss.who == us) {
+			Err(Error::AlreadySubmitted)
+		} else {
+			Ok(())
+		}
+	})
+}
+
+macro_rules! monitor_cmd_for { ($runtime:tt) => { paste::paste! {
+	/// The monitor command.
+	pub(crate) async fn [<monitor_cmd_ $runtime>](
+		client: WsClient,
+		shared: SharedConfig,
+		config: MonitorConfig,
+		signer: Signer,
+	) -> Result<(), Error> {
+		use $crate::[<$runtime _runtime_exports>]::*;
+		let (sub, unsub) = if config.listen == "head" {
+			("chain_subscribeNewHeads", "chain_unsubscribeNewHeads")
+		} else {
+			("chain_subscribeFinalizedHeads", "chain_unsubscribeFinalizedHeads")
+		};
+
+		log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", sub, unsub);
+		let mut subscription: Subscription<Header> = client
+			.subscribe(&sub, JsonRpcParams::NoParams, &unsub)
+			.await
+			.unwrap();
+
+		while let Ok(Some(now)) = subscription.next().await {
+			let hash = now.hash();
+			log::debug!(target: LOG_TARGET, "new event at #{:?} ({:?})", now.number, hash);
```

Comment:
```suggestion
			log::debug!(target: LOG_TARGET, "new subscription event at #{:?} ({:?})", now.number, hash);
```
just not to be confused with a frame `Event`